assets/src/editor/events/ticket-assignments-manager/hooks/use-assignments-calculator.js

```diff
@@ -0,0 +1,148 @@
+
+		switch ( true ) {
+			case ( dateEntities.length === 1 && datesMissing ) ||
+			( ticketEntities.length === 1 && datesMissing ) :
+				message = __(
+					'Event Dates must always have at least one Ticket assigned to them. If the current assignment is not correct, assign the correct Ticket first, then remove others as required.',
+					'event_espresso'
+				);
+				break;
+			case ( ticketEntities.length === 1 && bothMissing ) ||
+			( dateEntities.length === 1 && ticketsMissing ) :
+				message = __(
+					'Tickets must always have at least one Event Date assigned to them. If the current assignment is not correct, assign the correct Event Date first, then remove others as required.',
+					'event_espresso'
+				);
+				break;
+			case ( ticketsMissing ):
+				message = __(
+					'Tickets must always have at least one date assigned to them but one or more of the tickets below does not have any. Please correct the assignments for the highlighted cells.',
+					'event_espresso'
+				);
+				break;
+			default:
+				message = __(
+					'Event Dates must always have at least one Ticket assigned to them but one or more of the Event Dates below does not have any. Please correct the assignments for the highlighted cells.',
+					'event_espresso'
+				);
+		}
+		return message;
+	}, [ dateEntities, ticketEntities ] );
+
+	const [
+		hasMissingAssignments,
+		noAssignmentsMessage,
+		updatedAssignmentsCounts,
+	] = useMemo( () => {
+		let totalMissingAssignments = 0;
+		const missingAssignmentCounts = { dates: 0, tickets: 0 };
+		const updatedCounts = cloneDeep( assignedCounts );
+
+		// callback for counting assignments being added or removed
+		const countAssignmentChangesForDates = ( emptyCount, count, dateId ) => {
+
+			// first consider any assignedState.
+			const assignedStateForDate = assignedState.assigned[ dateId ];
+
+			if ( assignedStateForDate && Array.isArray( assignedStateForDate ) ) {
+				assignedStateForDate.forEach( ( ticketId ) => {
+					updatedCounts.dates[ dateId ]++;
+					updatedCounts.tickets[ ticketId ]++;
+				} );
+			}
+
+			// consider any removedState
+			const removedStateForDate = assignedState.removed[ dateId ];
+
+			if ( removedStateForDate && Array.isArray( removedStateForDate ) ) {
+				removedStateForDate.forEach( ( ticketId ) => {
+					updatedCounts.dates[ dateId ]--;
+					updatedCounts.tickets[ ticketId ]--;
+				} );
+			}
+
+			if ( updatedCounts.dates[ dateId ] === 0 ) {
+				missingAssignmentCounts.dates++;
+				emptyCount++;
+			}
+
+			return emptyCount;
+		};
+
+		const countAssignmentChangesForTickets = ( emptyCount, count ) => {
+			if ( count === 0 ) {
+				missingAssignmentCounts.tickets++;
+				emptyCount++;
+			}
+			return emptyCount;
+		};
+		totalMissingAssignments = reduce( updatedCounts.dates, countAssignmentChangesForDates, totalMissingAssignments );
+		totalMissingAssignments = reduce( updatedCounts.tickets, countAssignmentChangesForTickets, totalMissingAssignments );
```

Comment:
Two reducers are needed here because it's possible there may be tickets not assigned to any dates (or in any assigned state change), so we need to consider those for possible missing assignment counts.